flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/filesystem/FileSystemTableSource.java

```diff
@@ -314,20 +394,86 @@ public String asSummaryString() {
         return map;
     }
 
-    private int[] readFields() {
-        return projectedFields == null
-                ? IntStream.range(0, DataType.getFieldCount(getPhysicalDataType())).toArray()
-                : Arrays.stream(projectedFields).mapToInt(array -> array[0]).toArray();
+    // --------------------------------------------------------------------------------------------
+    // Methods to apply projections and metadata,
+    // will influence the final output and physical type used by formats
+    // --------------------------------------------------------------------------------------------
+
+    @Override
+    public void applyProjection(int[][] projectedFields, DataType producedDataType) {
+        this.projectFields = projectedFields;
+        this.producedDataType = producedDataType;
+    }
+
+    @Override
+    public void applyReadableMetadata(List<String> metadataKeys, DataType producedDataType) {
+        if (!metadataKeys.isEmpty()) {
```

Comment:
this could be problematic. the optimizer might call this method twice. once with the declared metadata columns and once with the actual used ones (maybe empty). It should be safe to remove this if branch?